keps/168-pending-workloads-visibility/README.md

```diff
@@ -0,0 +1,694 @@
+-->
+
+### Risks and Mitigations
+
+#### Too large objects
+
+As the number of pending workloads is arbitrarily large there is a risk that the
+status information about the workloads may exceed the etcd limit of 1.5Mi on
+object size.
+
+Exceeding the etcd limit has a risk that the LocalQueue controller updates can
+fail.
+
+In order to mitigate this risk we introduce the `MaxCount` configuration
+parameter to limit the maximal number of pending workloads in the status.
+Additionally, limit the maximal value of the parameter to 4000, see
+also [Choosing the limits and defaults for MaxCount](#choosing-the-limits-and-defaults-for-maxcount).
+
+We should also note that large queue objects might be problematic for the
+kubernetes API server, even if the etcd limit is not exceeded. For example,
+when there are many LocalQueue instances with watches, because in that case
+the entire LocalQueue objects need to be sent though the watch channels.
+
+To mitigate this risk we also extend the Kueue's user-facing documentation to
+warn about setting this number high on clusters with many LocalQueue instances,
+especially, when watches on the objects are used.
+
+#### Status updates for pending workloads slowing down other operations
+
+The operation of computing and updating the list of top pending workloads can
+have a degrading impact on the overall performance of other Kueue operations.
+
+This risk exists because the operation requires iteration over the contents of
+the cluster queue, which requires a read lock on the queue. Also, positional
+changes to the list of pending workloads may require more frequent updates if
+attempt to keep the information up-to-date.
+
+In order to mitigate the risk we maintain the statuses on best-effort basis,
+and issue at most one update request in a configured interval,
+see [throttling of status updates](#throttling-of-status-updates).
+
+Additionally, we take periodically an in-memory snapshot of the ClusterQueue to
+allow generation of the status with `MaxCount` elements for LocalQueues and
+ClusterQueues without taking the read lock for a prolonged time:
+[In-memory snapshot of the ClusterQueue](#In-memory snapshot of the ClusterQueue).
+
+#### Large number of API requests triggered after workload admissions
+
+In a scenario when we have multiple LocalQueues pointing to the same
+ClusterQueue a workload that is admitted in one LocalQueue shifts positions of
+pending workloads in other LocalQueues. In the worst case scenario updating the
+LocalQueue statuses with new positions requires as many API requests as the
+number of LocalQueues. In particular, sending over 100 requests after workload
+admission would degrade Kueue performance.
+
+First, we propose to batch the LocalQueue updates by time intervals. This helps
+to avoid sending API requests per LocalQueue if the positions are shifted
+multiple times in a short period of time.
+
+Second, we introduce the `MaxPosition` parameter. With this parameter, the
```

Comment:
In the Configuration API, right? Can you add it in the API definition below?

Reply:
Right, fixed, it is in the `Configuration API` section now.